Library/OcDeviceMiscLib/ForgeUefi.c

```diff
@@ -93,14 +110,34 @@ OcForgeUefiSupport (
     return EFI_INVALID_PARAMETER;
   }
 
-  NewBS = AllocateZeroPool (sizeof (EFI_BOOT_SERVICES));
-  if (NewBS == NULL) {
-    DEBUG ((DEBUG_INFO, "OCDM: Failed to allocate BS copy\n"));
-    return EFI_OUT_OF_RESOURCES;
+  if (Trash) {
+    Signature = *(UINT64 *)(&gBS->CreateEventEx);
+    if (Signature != DXE_SERVICES_SIGNATURE) {
+      DEBUG ((
+        DEBUG_INFO,
+        "OCDM: Aborting trash strategy 0x%016lX !=  0x%016lX\n",
+        Signature,
+        DXE_SERVICES_SIGNATURE
+        ));
+      return EFI_UNSUPPORTED;
+    }
+
+    DEBUG ((
+      DEBUG_INFO,
+      "OCDM: DXE signature 0x%016lX found, trashing for CreateEventEx\n",
+      Signature
+      ));
+    NewBS = gBS;
+  } else {
+    NewBS = AllocateZeroPool (sizeof (EFI_BOOT_SERVICES));
+    if (NewBS == NULL) {
+      DEBUG ((DEBUG_INFO, "OCDM: Failed to allocate BS copy\n"));
+      return EFI_OUT_OF_RESOURCES;
+    }
+
+    CopyMem (NewBS, gBS, gBS->Hdr.HeaderSize);
```

Comment:
What if `gBS->Hdr.HeaderSize != sizeof (EFI_BOOT_SERVICES)` (perfectly legal when spec versions of the host and our driver mismatch)?

Reply:
I believe we've checked this above, in line 109.

Reply:
Ohh, I missed `CreateEventEx` is actually populated below. Ok.